File: src/cam_transformer.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <ctime>
#include <cstdlib>
#include <string>
#include <dynamic_reconfigure/server.h>
#include <simple_camera_transformer/TransformerConfig.h>
#include <vector>

// Save the config state
simple_camera_transformer::TransformerConfig config_state;

// Publisher
image_transport::Publisher pub;

dynamic_reconfigure::Server<simple_camera_transformer::TransformerConfig> *server = nullptr;

void configCallback(simple_camera_transformer::TransformerConfig& config_, uint32_t level)
{
    config_state = config_;
}

void applyTransform(cv::Mat& frame, const simple_camera_transformer::TransformerConfig& config)
{
    // Resize
    cv::resize(frame, frame, cv::Size(), config.resize, config.resize, cv::INTER_AREA);

    // Color correction (brightness & contrast)
    if(config.enable_color_correct)
    {
        frame = frame*config.cc_alpha + config.cc_beta;
    }

    // Sharpen
    if(config.enable_sharpen)
    {
        cv::Mat out;
        cv::GaussianBlur(frame, out, cv::Size(0, 0), config.sharp_kernel*2+1);
        cv::addWeighted(frame, 1.0+config.sharp_weight, out, -1.0*config.sharp_weight, 0, out);
        out.copyTo(frame);
    }
}

void imageCallback(const sensor_msgs::ImageConstPtr &input)
{
    // Get the OpenCV image from the ROS message
    cv_bridge::CvImagePtr cv_ptr;
    try
    {
        cv_ptr = cv_bridge::toCvCopy(input, sensor_msgs::image_encodings::BGR8);
    }
    catch (cv_bridge::Exception &e)
    {
        // Something went wrong
        ROS_ERROR("cv_bridge exceptions: %s", e.what());
        return;
    }

    sensor_msgs::ImagePtr msg;
    cv::Mat frame = cv_ptr->image;

    // Apply transformations
    applyTransform(frame, config_state);

    // Convert it back to a message and publish
    msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();
    if(pub.getNumSubscribers() > 0)
    {
        pub.publish(msg);
    }

    cv::waitKey(3);
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "cam_transformer");

    ros::NodeHandle nh{"~"};
    image_transport::ImageTransport it(nh);

    dynamic_reconfigure::Server<simple_camera_transformer::TransformerConfig> server_;

    server = &server_;

    std::string pub_topic;
    std::string sub_topic;
    if (!nh.getParam("pub_topic", pub_topic) || !nh.getParam("sub_topic", sub_topic))
    {
        ROS_ERROR_STREAM("Missing 'pub_topic' or 'sub_topic' rosparams");
        return 0;
    }

    pub = it.advertise(pub_topic, 1);
    image_transport::Subscriber img_input = it.subscribe(sub_topic, 1, &imageCallback, 0);

    server_.setCallback(boost::bind(&configCallback, _1, _2));
    server_.getConfigDefault(config_state);

    ros::spin();
    }

File: scripts/cam_transformer.py
#!/usr/bin/env python3
import rospy
import cv2
import pdb
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
from dynamic_reconfigure.server import Server
from simple_camera_transformer.cfg import TransformerConfig
import numpy as np

class image_transformer:
    def __init__(self):
        # Initization of the node, name_sub
        rospy.init_node('cam_transformer', anonymous=True)

        # Get rosparams
        if rospy.has_param('~sub_topic'):
            sub_topic=rospy.get_param('~sub_topic')      
        else:
            print("Missing sub_topic parameter")
            exit(-1)

        if rospy.has_param('~pub_topic'):
            pub_topic=rospy.get_param('~pub_topic')      
        else:
            print("Missing pub_topic parameter")
            exit(-1)
            
        # Initialize the CvBridge class
        self.bridge = CvBridge()

        # Setup config server
        self.config=None
        self.cfg_srv = Server(TransformerConfig, self.configCallback)

        # Setup callback function
        self.image_sub = rospy.Subscriber(sub_topic, Image, self.image_callback)

        # Setup publisher
        self.pub = rospy.Publisher(pub_topic, Image, queue_size=10)

        rospy.spin()

    def configCallback(self, cfg, level):
        self.config=cfg
        return cfg

    def applyTransform(self, frame):
        if self.config is None:
            return frame

        # Resize
        frame=cv2.resize(frame, (0,0), fx=self.config.resize, fy=self.config.resize, interpolation=cv2.INTER_AREA)

        # Color correction (brightness & contrast)
        if self.config.enable_color_correct:
            frame = (frame*self.config.cc_alpha + self.config.cc_beta).astype(np.uint8)

        # Sharpen
        if self.config.enable_sharpen:
            out=cv2.GaussianBlur(frame, (0,0), self.config.sharp_kernel*2+1)
            frame = cv2.addWeighted( frame, 1.0+self.config.sharp_weight, out, -1.0*self.config.sharp_weight, 0).astype(np.uint8)

        return frame

    def image_callback(self, img_msg):
        # Convert the ROS Image message to a CV2 Image
        try:
            cv_image = self.bridge.imgmsg_to_cv2(img_msg, "bgr8")
        except CvBridgeError as e:
            rospy.logerr("CvBridge Error: {0}".format(e))    
            return    
        frame = self.applyTransform(cv_image)
        self.pub.publish(self.bridge.cv2_to_imgmsg(frame,encoding='bgr8'))
            
if __name__ == '__main__':
    IT=image_transformer()
 
